101-data-factory-v2-blob-to-blob-copy/azuredeploy.json

```diff
@@ -1,209 +1,193 @@
-      }
-    },
-        "blobContainer": {
-            "type": "string",
-            "defaultValue":"GET-PREREQ-blobContainer",
-            "metadata": {
-      	        "description": "Name of the blob container in the Azure Storage account."
-      }
-    },
-        "inputBlobFolder": {
-            "type": "string",
-            "defaultValue":"GET-PREREQ-blobFolder",
-            "metadata": {
-      	        "description": "The folder in the blob container that has the input file."
-      }
-    },
-        "inputBlobFileName": {
-            "type": "string",
-            "defaultValue":"GET-PREREQ-blobFileName",
-            "metadata": {
-                "description": "Name of the input file/blob."
-      	    }
-	    }
-    },
-    "variables":{
-        "factoryId":"[concat('Microsoft.DataFactory/factories/', parameters('dataFactoryName'))]"
-    },
-    "resources":[
-        {  
-            "type":"Microsoft.DataFactory/factories",
-            "name":"[parameters('dataFactoryName')]",
-            "apiVersion":"2018-06-01",
-            "location":"[parameters('location')]",
-            "identity":{  
-                "type":"SystemAssigned"
-            },
-            "resources":[  
-                {  
-                    "type":"Microsoft.DataFactory/factories/linkedServices",
-                    "name":"[concat(parameters('dataFactoryName'), '/ArmtemplateStorageLinkedService')]",
-                    "apiVersion":"2018-06-01",
-                    "properties":{  
-                        "annotations":[  
+	"$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+	"contentVersion": "1.0.0.0",
+	"parameters": {
+		"dataFactoryName": {
+			"type": "string",
+			"defaultValue": "[concat('datafactory', uniqueString(resourceGroup().id))]",
+			"metadata": {
+				"description": "Data Factory Name"
+			}
+		},
+		"location": {
+			"type": "string",
+			"defaultValue": "[resourceGroup().location]",
+			"metadata": {
+				"description": "Location of the data factory. Currently, only East US, East US 2, and West Europe are supported."
+			}
+		},
+		"storageAccountName": {
+			"type": "string",
+			"defaultValue": "[concat('storage', uniqueString(resourceGroup().id))]",
+			"metadata": {
+				"description": "Name of the Azure storage account that contains the input/output data."
+			}
+		},
+		"blobContainer": {
+			"type": "string",
+			"defaultValue": "[concat('blob', uniqueString(resourceGroup().id))]",
+			"metadata": {
+				"description": "Name of the blob container in the Azure Storage account."
+			}
+		}
+	},
+	"variables": {
+		"factoryId": "[resourceId('Microsoft.DataFactory/factories/', parameters('dataFactoryName'))]",
+		"storageAccountId": "[resourceId('Microsoft.Storage/storageAccounts', parameters('storageAccountName'))]",
+		"storageLinkedService": "[concat(variables('factoryId'), '/linkedServices/ArmtemplateStorageLinkedService')]",
```

Comment:
remove concat() and use resourceId()

Reply:
When I do that, I get an error:

Deployment template validation failed: 'The template variable 'storageLinkedService' is not valid: Unable to evaluate template language function 'resourceId': function requires exactly one multi-segmented argument which must be resource type including resource provider namespace. Current function arguments '/subscriptions/ab39dbc6-61cd-47cb-832c-c2c7c55391e8/resourceGroups/ADFARMtest/providers/Microsoft.DataFactory/factories/datafactoryva7rlg4aofcs6,/linkedServices/ArmtemplateStorageLinkedService'. Please see https://aka.ms/arm-template-expressions/#resourceid for usage details.. Please see https://aka.ms/arm-template-expressions for usage details.'.

I've seen in other templates, having contact in the main variables is acceptable.

Can I ask why not with these?

Reply:
you need to set the segments properly:

"storageLinkedService": "[resourceId('Microsoft.DataFactory/factories/linkedServices',  parameters('dataFactoryName'), 'ArmtemplateStorageLinkedService')]"

Follow the same pattern with the others

Reply:
done. :)